### PR TITLE
Feat add qianfan and dashscope

### PR DESCRIPTION
**Pull request by better629**
**Features**
<!-- Clear and direct description of the submit features. -->
<!-- If it's a bug fix, please also paste the issue link. -->

- add baidu qianfan
- add alibaba dashscope
- update repair_llm new case
    
**Feature Docs**
<!-- The RFC, tutorial, or use cases about the feature if it's a pretty big update. If not, there is no need to fill. -->
`https://github.com/geekan/MetaGPT-docs/blob/main/src/en/guide/get_started/configuration.md#baidu-qianfan-api`

**Influence**
<!-- Tell me the impact of the new feature and I'll focus on it.  -->

**Result**
<!-- The screenshot/log of unittest/running result -->

logs
[qwen-plus-20240223.txt](https://github.com/geekan/MetaGPT/files/14382094/qwen-plus-20240223.txt)

UT
```
(metagpt) MacBook-Pro:MetaGPT xxx$ pytest tests/metagpt/provider/test_qianfan_api.py                                                                                                                                                                 
tests/metagpt/provider/test_qianfan_api.py .                                                                                                                                                                                     [100%]

========================================================================================================== 1 passed in 0.05s ===================================================================
```

```
(metagpt) MacBook-Pro:MetaGPT xxx$ pytest tests/metagpt/provider/test_dashscope_api.py                                                                                                                                                                                                                
tests/metagpt/provider/test_dashscope_api.py .                                                                                                                                                                                   [100%]

========================================================================================================== 1 passed in 0.04s ======================================
```

```
(metagpt_v2) MacBook-Pro:MetaGPT xxxx$ pytest tests/metagpt/utils/test_repair_llm_raw_output.py 
========================================================================================================= test session starts ==========================================================================================================                                                                                                                                                                                                                    

tests/metagpt/utils/test_repair_llm_raw_output.py .......                                                                                                                                                                        [100%]

========================================================================================================== 7 passed in 3.83s ===========================================================================================================
```

**Other**
<!-- Something else about this PR. -->

### CODE REVIEW

File: metagpt/utils/token_counter.py [hunk 1]
```diff
@@ -38,6 +38,88 @@
 }
 
 
+"""
+QianFan Token Price https://cloud.baidu.com/doc/WENXINWORKSHOP/s/hlrk4akp7#tokens%E5%90%8E%E4%BB%98%E8%B4%B9
+Due to QianFan has multi price strategies, we unify `Tokens post-payment` as a statistical method.
+"""
+QianFan_MODEL_TOKEN_COSTS = {
```
Comment (geekan) [inline metagpt/utils/token_counter.py:45]: Keep all caps for global variables

File: metagpt/utils/token_counter.py [hunk 2]
```diff
@@ -38,6 +38,88 @@
...
+    "CodeLlama-7B-Instruct": {"prompt": 0.00056, "completion": 0.00056},
+    "XuanYuan-70B-Chat-4bit": {"prompt": 0.0049, "completion": 0.0049},
+    "Qianfan-BLOOMZ-7B-compressed": {"prompt": 0.00056, "completion": 0.00056},
+    "Qianfan-Chinese-Llama-2-7B": {"prompt": 0.00056, "completion": 0.00056},
+    "Qianfan-Chinese-Llama-2-13B": {"prompt": 0.00084, "completion": 0.00084},
+    "ChatLaw": {"prompt": 0.0011, "completion": 0.0011},
+    "Yi-34B-Chat": {"prompt": 0.0, "completion": 0.0},
+}
+
+QianFan_EndPoint_TOKEN_COSTS = {
```
Comment (geekan) [inline metagpt/utils/token_counter.py:69]: ditto

File: metagpt/utils/token_counter.py [hunk 3]
```diff
@@ -38,6 +38,88 @@
...
+    "qianfan_chinese_llama_2_13b": QianFan_MODEL_TOKEN_COSTS["Qianfan-Chinese-Llama-2-13B"],
+    "chatlaw": QianFan_MODEL_TOKEN_COSTS["ChatLaw"],
+    "yi_34b_chat": QianFan_MODEL_TOKEN_COSTS["Yi-34B-Chat"],
+}
+
+"""
+DashScope Token price https://help.aliyun.com/zh/dashscope/developer-reference/tongyi-thousand-questions-metering-and-billing
+Different model has different detail page. Attention, some model are free for a limited time.
+"""
+DashScope_TOKEN_COSTS = {
```
Comment (geekan) [inline metagpt/utils/token_counter.py:97]: ditto

File: examples/llm_hello_world.py [hunk 4]
```diff
@@ -13,7 +13,18 @@
 
 async def main():
     llm = LLM()
-    logger.info(await llm.aask("hello world"))
+    # llm type check
+    id_ques = "what's your name"
```
Comment (geekan) [inline examples/llm_hello_world.py:17]: Perhaps question is a better var name

File: metagpt/provider/base_llm.py [hunk 5]
```diff
@@ -67,6 +68,28 @@ def _system_msgs(self, msgs: list[str]) -> list[dict[str, str]]:
...
+        calc_usage = self.config.calc_usage and local_calc_usage
+        model = model if model else self.model
+        usage = usage.model_dump() if isinstance(usage, BaseModel) else usage
+        if calc_usage and self.cost_manager:
+            try:
+                prompt_tokens = int(usage.get("prompt_tokens", 0))
+                completion_tokens = int(usage.get("completion_tokens", 0))
+                self.cost_manager.update_cost(prompt_tokens, completion_tokens, model)
+            except Exception as e:
+                logger.error(f"{self.__class__.__name__} updats costs failed! exp: {e}")
```
Comment (geekan) [inline metagpt/provider/base_llm.py:86]: updats -> updates

File: metagpt/provider/qianfan_api.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        if self.use_system_prompt:
+            # if the model support system prompt, extract and pass it
+            if messages[0]["role"] == "system":
+                kwargs["messages"] = messages[1:]
+                kwargs["system"] = messages[0]["content"]  # set system prompt here
+        return kwargs
+
+    def _update_costs(self, usage: dict):
+        """update each request's token cost"""
+        model_or_endpoint = self.config.model if self.config.model else self.config.endpoint
```
Comment (geekan) [inline metagpt/provider/qianfan_api.py:106]: a if a else b
=> 
a or b

File: metagpt/provider/base_llm.py [hunk 7]
```diff
@@ -67,6 +68,28 @@ def _system_msgs(self, msgs: list[str]) -> list[dict[str, str]]:
...
         return self._system_msg(self.system_prompt)
 
+    def _update_costs(self, usage: Union[dict, BaseModel], model: str = None, local_calc_usage: bool = True):
+        """update each request's token cost
+        Args:
+            model (str): model name or in some scenarios called endpoint
+            local_calc_usage (bool): some models don't calculate usage, it will overwrite LLMConfig.calc_usage
+        """
+        calc_usage = self.config.calc_usage and local_calc_usage
+        model = model if model else self.model
```
Comment (geekan) [inline metagpt/provider/base_llm.py:78]: a if a else b
=>
a or b

File: metagpt/provider/qianfan_api.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        local_calc_usage = True if model_or_endpoint in self.token_costs else False
```
Comment (geekan) [inline metagpt/provider/qianfan_api.py:107]: local_calc_usage = model_or_endpoint in self.token_costs

File: metagpt/provider/dashscope_api.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        return output.get("choices", [{}])[0].get("message", {}).get("content", "")
+
+    def completion(self, messages: list[dict]) -> GenerationOutput:
+        resp: GenerationResponse = self.aclient.call(**self._const_kwargs(messages, stream=False))
+        self._check_response(resp)
+
+        self._update_costs(dict(resp.usage))
+        return resp.output
+
+    async def _achat_completion(self, messages: list[dict]) -> GenerationOutput:
```
Comment (geekan) [inline metagpt/provider/dashscope_api.py:213]: acompletion_text requires this function to return str. It seems that this type does not match

Comment (better629) [reply metagpt/provider/dashscope_api.py:213]: `_achat_completion` then `get_choice_text`

**Review by geekan — COMMENTED** [review]
LGTM

**Pull request merged**